### PR TITLE
feat: add parameter to ProxyAgent options, which is type of authentication

### PR DESCRIPTION
**Pull request by oooshima**
Implements: add parameter to ProxyAgent options

Benefits: ProxyAgent can use a variety of authentication types.

Also see: About this issue https://github.com/nodejs/undici/issues/1704

### CODE REVIEW

**Review by KhafraDev — APPROVED** [review]
cc @RafaelGSS

**Review by ronag — APPROVED** [review]

File: lib/proxy-agent.js [hunk 1]
```diff
@@ -55,10 +55,15 @@ class ProxyAgent extends DispatcherBase {
     this[kProxyTls] = opts.proxyTls
     this[kProxyHeaders] = {}
 
+    // deprecated
     if (opts.auth) {
       this[kProxyHeaders]['proxy-authorization'] = `Basic ${opts.auth}`
     }
 
+    if (opts.token) {
```
Comment (RafaelGSS) [inline lib/proxy-agent.js:63]:
```suggestion
    } else if (opts.token) {
```

File: lib/proxy-agent.js [hunk 2]
```diff
@@ -55,10 +55,15 @@ class ProxyAgent extends DispatcherBase {
     this[kProxyTls] = opts.proxyTls
     this[kProxyHeaders] = {}
 
+    // deprecated
```
Comment (RafaelGSS) [inline lib/proxy-agent.js:58]:
```suggestion
    // @deprecated in favour of opts.token
```

File: types/proxy-agent.d.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ declare namespace ProxyAgent {
   export interface Options extends Agent.Options {
     uri: string;
     auth?: string;
```
Comment (RafaelGSS) [inline types/proxy-agent.d.ts:17]:
```suggestion
       /**
        * @deprecated use opts.token
        */
        auth?: string;
```

Comment (RafaelGSS) [reply types/proxy-agent.d.ts:17]: Fix the lint too

**Review by RafaelGSS — CHANGES_REQUESTED** [review]
Could you please include docs for this new property? Would be great to mention the deprecation of `auth` in the docs too.

**Comment by oooshima** [comment]
@RafaelGSS
Thank you!
I have fixed it based on your comments.
https://github.com/nodejs/undici/pull/1705/commits/b2c64c089f042cb8892dd2cc2f6efa6562e999f2

File: docs/api/ProxyAgent.md [hunk 4]
```diff
@@ -74,6 +76,26 @@ for await (const data of body) {
...
 ```
 
+#### Example - Basic Proxy Request with authentication
+
+```js
+import { setGlobalDispatcher, request, ProxyAgent } from 'undici';
+
+const proxyAgent = new ProxyAgent({
+  uri: 'my.proxy.server',
+  token 'Bearer xxxx'
```
Comment (cola119) [inline docs/api/ProxyAgent.md:86]:
```suggestion
  token: 'Bearer xxxx'
```

Comment (oooshima) [reply docs/api/ProxyAgent.md]: @cola119
Sorry! I have fixed it.
https://github.com/nodejs/undici/pull/1705/commits/fb7a9c43eb4ecc05f87f8805693d32ded2cf8e47

File: lib/proxy-agent.js [hunk 5]
```diff
@@ -55,8 +55,11 @@ class ProxyAgent extends DispatcherBase {
     this[kProxyTls] = opts.proxyTls
     this[kProxyHeaders] = {}
 
+    // @deprecated in favour of opts.token
     if (opts.auth) {
       this[kProxyHeaders]['proxy-authorization'] = `Basic ${opts.auth}`
+    } else if (opts.token) {
+      this[kProxyHeaders]['proxy-authorization'] = opts.token
```
Comment (ronag) [inline lib/proxy-agent.js:62]: Maybe throw an error if both are provided?

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you add a test for the types?

File: lib/proxy-agent.js (hunk 5 above)
Comment (oooshima) [reply lib/proxy-agent.js:64]: @ronag
I thought that would be more helpful too, so I fixed it.
https://github.com/nodejs/undici/pull/1705/commits/b989e96f3c8bc9f2bdc98d1152afcbe80c10daad

**Comment by oooshima** [comment]
@mcollina
I have added a test for the types.
https://github.com/nodejs/undici/pull/1705/commits/0ba193c9d1d87ae72bd5280bc30d376e0c967d25

**Review by ronag — APPROVED** [review]

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Review by cola119 — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**